### ISSUE TITLE
[css-overflow] add method to prevent elements from contributing to scrollable overflow

### ISSUE THREAD

**Issue by johannesodland** [w3c/csswg-drafts#8361]
https://www.w3.org/TR/2022/WD-css-overflow-3-20221231/#scrollable

Sometimes it's useful to prevent elements from contributing to the scrollable overflow area. In particular elements that exist for purely decorative purposes. This is mostly relevant for elements that have a used value of `position: fixed` or `position: absolute`.

For elements not participating in a 3D rendering context its possible to eliminate them from contributing to the scrollable overflow area by adding `contain: paint`, `overflow: hidden`, `overflow: clip` to a parent, effectively clipping them away.

This is not a workable strategy for elements participating in a 3D rendering context. `contain: paint`, `overflow: hidden` and `overflow: clip` are all grouping property values, and will force the parent element to have a used style of `flat` for `preserve-3d`. This effectively separates the elements from the 3d rendering context they were participating in. 

### Proposal
Add a method to prevent elements from contributing to scrollable overflow. 
This could be a new property such as `overflow-contribution: normal | none`.

### Example
```html
<div class="parallax-scroll-container">
  <div class="decorative-element">...</div>
  ...
</div>
```
```css
.parallax-scroll-container {
  perspective: 100px;
  overflow: auto;
}

/* 
This element will extend the scrollable overflow area, and clipping it 
will prevent it from participating in the perspective transformed 3d rendering context 
*/
.decorative-element {
  --depth: 1.5;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  height: 200vh; /* Arbitrary height taller than content */
  transform: translateZ(calc(-100px * (var(--depth) - 1)))
    scale(calc(var(--depth)));
  transform-origin: 50vw 50vh;
}
```

**Comment by jonjohnjohnson** [comment]
Could be worth keeping #3322 in mind here.

**Comment by LeaVerou** [comment]
Some more discussion in https://github.com/w3c/csswg-drafts/issues/8400, which is closed as a duplicate.

**Comment by dbaron** [comment]
Some more naming possibilities (some of which are from or partly from the discussion above; I didn't see any in #8400) include (listing the initial value first):
* overflow-contribution: normal | none
* overflow-skip: none | skip
* overflow-extend: normal | none

**Comment by bfgeek** [comment]
One thing that I've heard from developers is a desire just to skip the transform adjustment that happens. E.g. something that could be useful is:

`overflow-contribution: normal | auto | ignore-transform`

**Comment by dbaron** [comment]
(see also #9458 about defining exactly *what* transform adjustment happens!)

**Comment by LeaVerou** [comment]
> One thing that I've heard from developers is a desire just to skip the transform adjustment that happens. E.g. something that could be useful is:
> 
> `overflow-contribution: normal | none | ignore-transform`

If this is about elements being moved by transforms triggering overflow, it's not exclusive to transforms, it also applies to positioning.

**Comment by flackr** [comment]
If this property affects scrollable overflow, it should probably also affect scrolling related APIs and snap areas. E.g. I could see `overflow-contribution: ignore-transform` being really useful to:
* avoid the pitfalls of things like animating transform on `scroll-snap-align: center` element which currently can lead to jittery scroll updates.
* scroll to the correct position which follows an entry animation, e.g. a common entry animation is to transform and fade in. In combination with this property you could scroll to the location the element will be in when the animation finishes.

As such, maybe the name needs to be more generic like `overflow-box`?

**Comment by bfgeek** [comment]
Adgenda+ to paint the bikeshed on a name for this property.

**Comment by fantasai** [comment]
I think for the purpose of learnability, it would be better to have this as one property rather than two.

**Comment by flackr** [comment]
To summarize the discussion, there are two behaviors being discussed here:
1. Whether or not a given element contributes to the scrollable overflow.
2. What the used rectangle for scrollable overflow and scrolling operations for this element should be. E.g. should it ignore the element's transform.

I think we are all agreed that these are both valuable behaviors to be able to control. The main point of contention was whether this would be 1 or 2 properties. To help I'll suggest possible properties / values (of course welcome bikeshedding).

As one property, this would look something like this:
```css
overflow-contribution: normal | ignore-transform | none;
```
The question is if we specify none, what are the effects on scrolling / snapping operations? Would we prevent snapping / scrolling to this element? If not, what if you want the element to not contribute to scrollable overflow and ignore transforms for scroll related operations?
A counter-point to this, is when you want to not contribute to scrollable overflow, the UA won't guarantee that this element is reachable which may mean that the developer doesn't intend the user to need to scroll to this element, so maybe it would be reasonable to treat the element as purely decorative?

As two properties, these would be specified independently, e.g.:
```css
overflow-contribution: normal | none;
scroll-align: normal | ignore-transform;
```
Do all combinations of this make sense? Probably, you can imagine how these could be applied independently, but are there practical use cases where we want to scroll to an element that doesn't contribute to scrollable overflow? Perhaps an overly large image that you don't want to introduce horizontal scrolling for?

Alternately, a third option could be to have a single property with multiple values, but this may just suggest a need for multiple longhand properties, e.g.
```css
overflow-contribution: <normal | none> || <auto | ignore-transform>?;
```

**Comment by johannesodland** [comment]
> The question is if we specify none, what are the effects on scrolling / snapping operations? Would we prevent snapping / scrolling to this element?

Could we treat this similar to an element that is placed wholly or partly into the [negative scrollable overflow region](https://www.w3.org/TR/css-overflow-3/#negative-scrollable-overflow-region)? Such an element won't extend the scrollable overflow area, and the [used snap point](https://www.w3.org/TR/css-scroll-snap-1/#unreachable) for the element is the position resulting from scrolling as much as possible towards the desired snap point in the relevant axis?

Regarding snapping operations: there is a related issue https://github.com/w3c/csswg-drafts/issues/7885 containing use cases for extending the scrollable overflow area to make snap points reachable. Maybe that issue could be solved with the same property/-ies?

**Comment by johannesodland** [comment]
@flackr Any chance we can pick this issue up again?

**Comment by johannesodland** [comment]
> 2\. What the used rectangle for scrollable overflow and scrolling operations for this element should be. E.g. should it ignore the element's transform.

Regarding scrolling operations, it would be useful to be able to ignore the sticky offsets of an element. That way, scroll operations would scroll to the in-flow position of a sticky element. This would be easier if `scroll-align` was a separate property.

```css
scroll-align: normal | ignore-transform | ignore-sticky-offsets;
```

**Comment by flackr** [comment]
> Regarding scrolling operations, it would be useful to be able to ignore the sticky offsets of an element. That way, scroll operations would scroll to the in-flow position of a sticky element.

Yeah we tried to do this earlier by default (see #1459), however it turned out to have many cases where it wasn't reasonable behavior so perhaps having it as a controllable behavior would allow authors to select it when it made sense.

> This would be easier if `scroll-align` was a separate property.

I agree, sticky position is something that wouldn't make sense to ignore for overflow contribution but is very useful for scroll alignment.

In order to avoid making the property sound like it is doing the same thing as `scroll-snap-align` ([which can be used for scroll alignment without snapping](https://drafts.csswg.org/css-scroll-snap-1/#:~:text=snap%20container.-,The%20user%20agent%20may%20also%20do%20this%20even%20when%20the%20scroll%20container%20has%20scroll%2Dsnap%2Dtype%3A%20none.,-If%20a%20scroll)), it would probably be good to give it a name like `scroll-align-origin` or `scroll-align-box`.

TLDR, I propose two new properties:
```
scroll-align-origin: normal | ignore-transform || ignore-sticky-offset;
overflow-contribution: normal | none;
```
